deepchem/models/tensorgraph/models/seqtoseq.py

```diff
@@ -0,0 +1,267 @@
+
+    The prediction is done using a beam search with length normalization.
+
+    Parameters
+    ----------
+    sequence: sequence
+      the input sequence to generate a prediction for
+    beam_width: int
+      the beam width to use for searching.  Set to 1 to use a simple greedy search.
+    """
+    feed_dict = {}
+    feed_dict[self._features] = self._create_input_array([sequence])
+    feed_dict[self._gather_indices] = [(i, len(sequence))
+                                       for i in range(self.batch_size)]
+    feed_dict[self._training_placeholder] = 0.0
+    for initial, zero in zip(self.rnn_initial_states, self.rnn_zero_states):
+      feed_dict[initial] = zero
+    with self._get_tf("Graph").as_default():
+      probs = self.session.run(self.output, feed_dict=feed_dict)[0]
+    return self._beam_search(probs, beam_width)
+
+  def predict_from_embedding(self, embedding, beam_width=5):
+    """Given an embedding vector, predict the output sequence.
+
+    The prediction is done using a beam search with length normalization.
+
+    Parameters
+    ----------
+    embedding: array
+      the embedding vector to generate a prediction for
+    beam_width: int
+      the beam width to use for searching.  Set to 1 to use a simple greedy search.
+    """
+    feed_dict = {}
+    feed_dict[self.embedding] = embedding
+    feed_dict[self._training_placeholder] = 0.0
+    for initial, zero in zip(self.rnn_initial_states, self.rnn_zero_states):
+      feed_dict[initial] = zero
+    with self._get_tf("Graph").as_default():
+      probs = self.session.run(self.output, feed_dict=feed_dict)[0]
+    return self._beam_search(probs, beam_width)
+
+  def predict_embedding(self, sequence):
+    """Given an input sequence, compute the embedding vector.
+
+    Parameters
+    ----------
+    sequence: sequence
+      the input sequence to generate an embedding vector for
+    """
+    feed_dict = {}
+    feed_dict[self._features] = self._create_input_array([sequence])
+    feed_dict[self._gather_indices] = [(i, len(sequence))
+                                       for i in range(self.batch_size)]
+    feed_dict[self._training_placeholder] = 0.0
+    for initial, zero in zip(self.rnn_initial_states, self.rnn_zero_states):
+      feed_dict[initial] = zero
+    with self._get_tf("Graph").as_default():
+      return self.session.run(self.embedding, feed_dict=feed_dict)
+
+  def _beam_search(self, probs, beam_width):
+    """Perform a beam search for the most likely output sequence."""
+    if beam_width == 1:
+      # Do a simple greedy search.
+
+      s = []
+      for i in range(len(probs)):
+        token = self._output_tokens[np.argmax(probs[i])]
+        if token == SeqToSeq.sequence_end:
+          break
+        s.append(token)
+      return s
+
+    # Do a beam search with length normalization.
+
+    logprobs = np.log(probs)
+    # Represent each candidate as (normalized prob, raw prob, sequence)
+    candidates = [(0.0, 0.0, [])]
```

Comment:
maybe move this into a different function?  Looking through the most likely beams and their normalized prob is super useful.

Also impressive implementation.